### PR TITLE
Damcilva/3.0/tools/signatures check pr gate fix no spec

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [ ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

The source file hash checker PR failed on some edge cases:
1) If a directory `./SPECS*/<dir>` did not contain a `*.spec` file the packer would fail (i.e., `SPECS/LICENSES-AND-NOTICES/*`)
2) If a directory had multiple `*.spec` files inside it, it would only pick the one that matched the directory name
3) If the `*.spec` file inside the directory had a different name than the directory.

###### Change Log  

- Handle edge cases in check-source-signatures.yml

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- This PR has test data included currently

### CODE REVIEW

File: .github/workflows/check-source-signatures.yml [hunk 1]
```diff
@@ -45,9 +45,25 @@ jobs:
...
+          # For each package directory, get the names of all .spec files contained and add them to the list of SRPMS to repack.
+          # We need to filter since some directories may not contain .spec files we can rebuild.
+          changed_pkgs=""
+          for pkg in $changed_pkg_dirs; do
+            changed_pkgs="$changed_pkgs $(find SPECS/$pkg -name '*.spec' -exec basename {} \; | sed 's/\.spec//g' | xargs)"
+          done
+          changed_pkgs_extended=""
+          for pkg in $changed_pkg_dirs_extended; do
+            changed_pkgs_extended="$changed_pkgs_extended $(find SPECS-EXTENDED/$pkg -name '*.spec' -exec basename {} \; | sed 's/\.spec//g' | xargs)"
+          done
```
Comment (PawelWMS) [inline .github/workflows/check-source-signatures.yml:65]: An idea to shorten things a bit:
```suggestion
          # the package's directory name (ie the folder inside ./SPECS).
          changed_pkg_dirs=$(git diff-tree --diff-filter=d --no-commit-id --name-only -r ${{ env.base_sha }} ${{ github.sha }} | { grep -oP "SPECS[^/]*/[^/]+" || [[ $? -eq 1 ]]; } | sort -u)

          echo "Folders with modified files in this PR:"
          echo "${changed_pkg_dirs}"

          # For each package directory, get the names of all .spec files contained and add them to the list of SRPMS to repack.
          # We need to filter since some directories may not contain .spec files we can rebuild.
          changed_pkgs=()
          while IFS= read -r pkg_dir; do
            changed_pkgs+=("$(find $pkg_dir -name '*.spec' -exec basename {} .spec \;)")
          done <<< "$changed_pkg_dirs"
```

You'd then also need to change

`echo "changed_pkgs=${changed_pkgs[@]}" >> $GITHUB_ENV`
to
`echo "changed_pkgs=${changed_pkgs}" >> $GITHUB_ENV.

**Comment by dmcilvaney** [comment]
https://github.com/microsoft/azurelinux/actions/runs/8486248404/job/23252362863?pr=8614 check with expected fail

File: .github/workflows/check-source-signatures.yml [hunk 2]
```diff
@@ -45,14 +48,22 @@ jobs:
       - name: Get changed packages
         run: |
           # Find the packages that have been modified in the current PR. They will be of the form '/path/to/SPECS/<pkgname>/**/.*', and we want to extract
-          # the package name (ie the folder inside ./SPECS).
-          changed_pkgs=$(git diff-tree --diff-filter=d --no-commit-id --name-only -r ${{ env.base_sha }} ${{ github.sha }} | { grep "SPECS/.*" || test $? = 1; } | sed -n 's#SPECS/\([^/]*\)/.*#\1#p' | sort -u | xargs)
-          changed_pkgs_extended=$(git diff-tree --diff-filter=d --no-commit-id --name-only -r ${{ env.base_sha }} ${{ github.sha }} | { grep "SPECS-EXTENDED/.*" || test $? = 1; } | sed -n 's#SPECS-EXTENDED/\([^/]*\)/.*#\1#p' | sort -u | xargs)
+          # the package's directory name (ie the folder inside ./SPECS).
+          changed_pkg_dirs=$(git diff-tree --diff-filter=d --no-commit-id --name-only -r ${{ env.base_sha }} ${{ github.sha }} | { grep "${{ matrix.specs-dir }}/.*" || test $? = 1; } | sed -n 's#SPECS/\([^/]*\)/.*#\1#p' | sort -u | xargs)
```
Comment (PawelWMS) [inline .github/workflows/check-source-signatures.yml:52]:
```suggestion
          changed_pkg_dirs=$(git diff-tree --diff-filter=d --no-commit-id --name-only -r ${{ env.base_sha }} ${{ github.sha }} | { grep "${{ matrix.specs-dir }}/.*" || test $? = 1; } | sed -n 's#${{ matrix.specs-dir }}/\([^/]*\)/.*#\1#p' | sort -u | xargs)
```

I still like the `grep -oP` version better to drop `sed` but as long as it works, I'm good.:)

File: .github/workflows/check-source-signatures.yml [hunk 3]
```diff
@@ -72,44 +83,27 @@ jobs:
...
         run: |
-          if [ -z "${{ env.changed_pkgs }}" ] && [ -z "${{ env.changed_pkgs_extended }}" ]; then
-            echo "No package changes detected."
+          if [ -z "${{ env.changed_pkgs }}" ]; then
+            echo "No package changes detected in '${{ matrix.specs-dir }}''."
             exit 0
           fi
 
-          # Core SPECs
           if [ -n "${{ env.changed_pkgs }}" ]; then
```
Comment (PawelWMS) [inline .github/workflows/check-source-signatures.yml:91]: This `if` shouldn't be needed any more.

File: .github/workflows/check-source-signatures.yml [hunk 4]
```diff
@@ -72,44 +83,27 @@ jobs:
...
-            sudo make -C toolkit -j$(nproc) input-srpms REBUILD_TOOLS=y DAILY_BUILD_ID=${{ env.LKG_BUILD_ID }} SRPM_PACK_LIST="${{ env.changed_pkgs }}"
-            core_err=$?
-            set +x
-            set -e
-          fi
-
-          # Extended SPECs
-          if [ -n "${{ env.changed_pkgs_extended }}" ]; then
-            # We want to ignore errors here, as we want to check all the packages that have been modified. Capture the error code and check it later.
+            # We want to ignore errors here so we can print a helpful message at the end.
```
Comment (PawelWMS) [inline .github/workflows/check-source-signatures.yml:92]: Idea: if you want to drop disabling/enabling `-e` and the `err` variable, you can also go for:
```bash
if ! sudo make -C toolkit (...); then
  printf "\n\n******************************"
...
```

File: .github/workflows/check-source-signatures.yml (hunk 2 above)
Comment (dmcilvaney) [reply .github/workflows/check-source-signatures.yml:52]: Oh, I lost the grep in a fixup. I didn't mind it either. I'll put it back.

Comment (dmcilvaney) [reply .github/workflows/check-source-signatures.yml:52]: Oh, I remember why I removed it now... `grep -oP "(?<=${{ matrix.specs_folder }}/)([^/]+)"` But I've put it back because I hate reading `sed`.

File: .github/workflows/check-source-signatures.yml (hunk 4 above)
Comment (dmcilvaney) [reply .github/workflows/check-source-signatures.yml:92]: I think the `set -x` will produce messy output if its outside the conditional, I'll probably leave it like this.

File: .github/workflows/check-source-signatures.yml [hunk 5]
```diff
@@ -45,14 +48,22 @@ jobs:
       - name: Get changed packages
         run: |
           # Find the packages that have been modified in the current PR. They will be of the form '/path/to/SPECS/<pkgname>/**/.*', and we want to extract
-          # the package name (ie the folder inside ./SPECS).
-          changed_pkgs=$(git diff-tree --diff-filter=d --no-commit-id --name-only -r ${{ env.base_sha }} ${{ github.sha }} | { grep "SPECS/.*" || test $? = 1; } | sed -n 's#SPECS/\([^/]*\)/.*#\1#p' | sort -u | xargs)
-          changed_pkgs_extended=$(git diff-tree --diff-filter=d --no-commit-id --name-only -r ${{ env.base_sha }} ${{ github.sha }} | { grep "SPECS-EXTENDED/.*" || test $? = 1; } | sed -n 's#SPECS-EXTENDED/\([^/]*\)/.*#\1#p' | sort -u | xargs)
+          # the package's directory name (ie the folder inside ./SPECS).
+          changed_pkg_dirs=$(git diff-tree --diff-filter=d --no-commit-id --name-only -r ${{ env.base_sha }} ${{ github.sha }} | { grep -oP "(?<=${{ matrix.specs-dir }}/)([^/]+)" || [[ $? -eq 1 ]]; } | sort -u | xargs)
```
Comment (PawelWMS) [inline .github/workflows/check-source-signatures.yml:52]: One thing I just noticed after people started complaining the PR check failed when I moved the `LICENSES-AND-NOTICES` folder:

```suggestion
          changed_pkg_dirs=$(git diff-tree --diff-filter=d --no-commit-id --name-only -r ${{ env.base_sha }} ${{ github.sha }} | { grep -oP "(?<=^${{ matrix.specs-dir }}/)([^/]+)" || [[ $? -eq 1 ]]; } | sort -u | xargs)
```

File: .github/workflows/check-source-signatures.yml (hunk 4 above)
Comment (dmcilvaney) [reply .github/workflows/check-source-signatures.yml]: Moved to using the conditional since the output wasn't formatting nicely the other way

File: .github/workflows/check-source-signatures.yml (hunk 5 above)
Comment (dmcilvaney) [reply .github/workflows/check-source-signatures.yml]: Fixed
